### ISSUE TITLE
重新训练的模型不能用

### ISSUE THREAD

**Issue by galaxyGGG** [csuhan/s2anet#79]
为了走一遍流程，我用一小部分dota数据进行了训练，但是用训练好的模型做inference的时候报错，请问是什么原因？

用自己的数据集也不行，我修改了num_classes，但是还是不行。

![2021-05-12 15-45-07屏幕截图](https://user-images.githubusercontent.com/45815385/119429491-882b6000-bd41-11eb-9bff-d86dacdbf88f.png)

**Comment by csuhan** [comment]
可能是训练效果较差，导致`anchors[topk_inds,:]`中`topk_inds`均为False

**Comment by galaxyGGG** [comment]
> 可能是训练效果较差，导致`anchors[topk_inds,:]`中`topk_inds`均为False

是不是因为我的数据集太少了？我为了让训练快点结束，只用了几张原图做训练集。训练的时候后面几个loss全是nan，请问这是什么问题？

![Screenshot from 2021-05-26 09-50-10](https://user-images.githubusercontent.com/45815385/119590963-54ffd400-be08-11eb-8697-f96e9160386b.png)

如果是数据太少的话，那么需要多少数据才合适呢？谢谢！

**Comment by csuhan** [comment]
学习率可能太大了，我们的配置都是4卡的

**Comment by galaxyGGG** [comment]
@csuhan 
我按照说明把学习率改小为0.00125(单卡 * 1 img/gpu),甚至改到0，loss都还是nan

config文件如下：

# model settings
model = dict(
    type='S2ANetDetector',
    #pretrained='torchvision://resnet50',
    pretrained = "///data/checkpoints/resnet50-19c8e357.pth",
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs=True,
        num_outs=5),
    bbox_head=dict(
        type='S2ANetHead',
        num_classes=16,
        in_channels=256,
        feat_channels=256,
        stacked_convs=2,
        with_orconv=True,
        anchor_ratios=[1.0],
        anchor_strides=[8, 16, 32, 64, 128],
        anchor_scales=[4],
        target_means=[.0, .0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0, 1.0],
        loss_fam_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_fam_bbox=dict(
            type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0),
        loss_odm_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_odm_bbox=dict(
            type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0)))
# training and testing settings
train_cfg = dict(
    fam_cfg=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.4,
            min_pos_iou=0,
            ignore_iof_thr=-1,
            iou_calculator=dict(type='BboxOverlaps2D_rotated')),
        bbox_coder=dict(type='DeltaXYWHABBoxCoder',
                        target_means=(0., 0., 0., 0., 0.),
                        target_stds=(1., 1., 1., 1., 1.),
                        clip_border=True),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    odm_cfg=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.4,
            min_pos_iou=0,
            ignore_iof_thr=-1,
            iou_calculator=dict(type='BboxOverlaps2D_rotated')),
        bbox_coder=dict(type='DeltaXYWHABBoxCoder',
                        target_means=(0., 0., 0., 0., 0.),
                        target_stds=(1., 1., 1., 1., 1.),
                        clip_border=True),
        allowed_border=-1,
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    nms_pre=2000,
    min_bbox_size=0,
    score_thr=0.05,
    nms=dict(type='nms_rotated', iou_thr=0.1),
    max_per_img=2000)
# dataset settings
dataset_type = 'DotaDataset'
data_root = '/data/dota_1024/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='RotatedResize', img_scale=(1024, 1024), keep_ratio=True),
    dict(type='RotatedRandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1024, 1024),
        flip=False,
        transforms=[
            dict(type='RotatedResize', img_scale=(1024, 1024), keep_ratio=True),
            dict(type='RotatedRandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    imgs_per_gpu=1,
    workers_per_gpu=1,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'trainval_split/trainval_s2anet.pkl',
        img_prefix=data_root + 'trainval_split/images/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'trainval_split/trainval_s2anet.pkl',
        img_prefix=data_root + 'trainval_split/images/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'test_split/test_s2anet.pkl',
        img_prefix=data_root + 'test_split/images/',
        pipeline=test_pipeline))
evaluation = dict(
    gt_dir='data/dota/test/labelTxt/', # change it to valset for offline validation
    imagesetfile='data/dota/test/test.txt')
# optimizer
optimizer = dict(type='SGD', lr=0, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[8, 11])
checkpoint_config = dict(interval=1)
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
    ])
# runtime settings
total_epochs = 12
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]

**Comment by galaxyGGG** [comment]
@csuhan 
我好像知道了，我是在docker环境跑的程序，应该是配的环境出问题了，谢谢！

**Comment by soHardToHaveAName** [comment]
> @csuhan 我好像知道了，我是在docker环境跑的程序，应该是配的环境出问题了，谢谢！

请问这个问题解决了吗？

**Comment by galaxyGGG** [comment]
解决了
















在 2022-03-26 21:36:44，"wang chong yu" ***@***.***> 写道：

@csuhan 我好像知道了，我是在docker环境跑的程序，应该是配的环境出问题了，谢谢！

请问这个问题解决了吗？

**Comment by soHardToHaveAName** [comment]
> 解决了 在 2022-03-26 21:36:44，"wang chong yu" ***@***.***> 写道： @csuhan 我好像知道了，我是在docker环境跑的程序，应该是配的环境出问题了，谢谢！ 请问这个问题解决了吗？ — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you modified the open/close state.Message ID: ***@***.***>

我是用的作者给的docker跑的，也出现了这个问题。是不是得自己从头配环境才行阿？

**Comment by soHardToHaveAName** [comment]
> > 解决了 在 2022-03-26 21:36:44，"wang chong yu" _**@**_._**> 写道： @csuhan 我好像知道了，我是在docker环境跑的程序，应该是配的环境出问题了，谢谢！ 请问这个问题解决了吗？ — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you modified the open/close state.Message ID: _**@**_.**_>
> 
> 我是用的作者给的docker跑的，也出现了这个问题。是不是得自己从头配环境才行阿？

后续：已解决。不使用作者提供的docker，在本地从头搭建环境，loss变得正常。
